pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -222,6 +223,11 @@ class ModelInterpreter implements Serializable {
 
     }
 
+    @NonCPS
+    private String getFlowNodeId() {
+        return script.getContext(FlowNode.class).getId()
```

Comment:
Can we not use the `withContext` step? It is intended for use in (really) power user trusted libraries, and it is just going to add overhead to the flow node graph for every Declarative build (as well as breaking tons of tests).

Would be better to add some helper method to `Utils` that is run silently, without recording a flow node.

Reply:
> Can we not use the withContext step?

`getContext`

> It is intended for use in (really) power user trusted libraries

What is `ModelInterpreter.groovy` if not an incredibly complicated trusted library written by power users?

> and it is just going to add overhead to the flow node graph for every Declarative build

One extra `FlowNode` per skipped stage. I do not think this is a big problem.

> as well as breaking tons of tests

It breaks 6 known tests across OSS plugins and CloudBees-internal plugins, and PRs have already been filed to fix all of them.

> Would be better to add some helper method to Utils that is run silently, without recording a flow node.

Yes, but here is the problem: Declarative has only a stage name, and wants to add/update tags on the corresponding `FlowNode` objects. Without an initial `FlowNode` to guide the scan, we have to scan the whole `FlowGraph`, and I am suspicious that there are some unfixable bugs with that approach given duplicate stage names in complex Pipelines. #698 shows that kind of approach. I think it is much clearer to just get a contextual ID to guide the search and the tradeoffs of having to use `getContext` do not seem that severe.

Perhaps the code could be adjusted to check the current heads and guess which one is the right one, but I think that is still problematic. For example if you use `matrix` and have two identically named stages/parallel branches, but then only one of them gets skipped, it does not seem trivial to figure out which is the "right" one.

Reply:
> Perhaps the code could be adjusted to check the current heads and guess which one is the right one, but I think that is still problematic. For example if you use matrix and have two identically named stages/parallel branches, but then only one of them gets skipped, it does not seem trivial to figure out which is the "right" one.

Maybe `CpsThread.head` could be used to avoid this problem, but I am not sure, particularly given its Javadoc.

Reply:
> Without an initial `FlowNode` to guide the scan

I see, this is the crux of the problem: there actually is no obvious Java/Groovy to call that would return usable information at this point. `CpsThread.current().etc()` would be tricky.